minecraft-common/src/main/java/com/sigmundgranaas/forgero/minecraft/common/toolhandler/PatternBreakingStrategy.java

```diff
@@ -6,85 +6,88 @@
-
-                        } else if (primary == Direction.NORTH || primary == Direction.SOUTH) {
-                            newPos = new BlockPos(rootPos.getX() + j - centerX, rootPos.getY() + i - centerY, rootPos.getZ());
-
-                        } else {
-                            Direction secondary = dir[1];
-                            if (secondary == Direction.EAST || secondary == Direction.WEST) {
-                                newPos = new BlockPos(rootPos.getX() + j - centerX, rootPos.getY(), rootPos.getZ());
-                            } else if (secondary == Direction.NORTH || secondary == Direction.SOUTH) {
-                                newPos = new BlockPos(rootPos.getX(), rootPos.getY(), rootPos.getZ() + i - centerY);
-                            } else {
-                                newPos = new BlockPos(rootPos.getX() + j - centerX, rootPos.getY(), rootPos.getZ() + i - centerY);
-                            }
-                        }
-                    } else {
-                        newPos = new BlockPos(rootPos.getX() + j - centerX, rootPos.getY(), rootPos.getZ() + j - centerX);
-                    }
-
-                    BlockState newState = world.getBlockState(newPos);
-                    if (isBreakableBlock(world, newPos, player) && (breakingPattern.checkBlock(newState) || newPos.equals(rootPos))) {
-                        list.add(new Pair<>(newState, newPos));
-                    }
-                }
-            }
-        }
-        return list;
-    }
+	private final PatternBreaking breakingPattern;
+
+	public PatternBreakingStrategy(PatternBreaking breakingPattern) {
+		this.breakingPattern = breakingPattern;
+	}
+
+	public PatternBreakingStrategy(PropertyData data) {
+		BreakingDirection dir = data.getDescription() == null ? BreakingDirection.ANY : data.getDirection();
+		if (data.getTags().size() > 0) {
+			this.breakingPattern = new TaggedPatternBreaking(data.getPattern(), dir, data.getTags().get(0));
+		} else {
+			this.breakingPattern = new PatternBreaking(data.getPattern(), dir);
+		}
+	}
+
+	@Override
+	public Set<BlockPos> getAvailableBlocks(BlockView world, BlockPos rootPos, PlayerEntity player) {
+		Direction[] directions = Direction.getEntityFacingOrder(player);
+		var availableBlocks = new HashSet<BlockPos>();
+
+		if (breakingPattern.getPattern().length == 0 || breakingPattern.getPattern()[0].length() % 2 == 0) {
+			return new HashSet<>();
+		}
+
+		int centerY = (breakingPattern.getPattern().length - 1) / 2;
+		if (breakingPattern.getPattern().length == 2) {
+			centerY = 1;
+		}
+		if (breakingPattern.getPattern().length == 1) {
+			centerY = 0;
+		}
+
+		int centerX = (breakingPattern.getPattern()[0].length() - 1) / 2;
+		if (breakingPattern.getPattern()[0].length() == 1) {
+			centerX = 0;
+		}
+
+		for (int i = 0; i < breakingPattern.getPattern().length; i++) {
```

Comment:
I'll add this in a separate PR next week, where I'll also look at further optimisations and improvements to other veinmining code.
This way we can get this merged sooner rather than later.